File: LeetCode/p53.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_res = nums[0]
        
        current = 0
        
        for i in nums:
            if current < 0:
                print("Reset", i)
                current = 0
                
            current += i
            max_res =  max(max_res, current)    
            print("current", i, current, max_res)
            
            
        return max_res



if __name__ == '__main__':
    sol = Solution()
    print(sol.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))

File: LeetCode-Java/p53.java
class Solution {
    public int maxSubArray(int[] nums) {
        int maxRes = nums[0];
        int current = 0;
        
        for (int num : nums) {
            if (current < 0) {
                current = 0;
            }
            current += num;
            maxRes = Math.max(maxRes, current);
        }
        
        return maxRes;
    }
}
